torchtitan/parallelisms/offload_utils.py

```diff
@@ -0,0 +1,136 @@
+
+
+class Handle:
+    def __init__(
+        self,
+        device_tensor: torch.Tensor,
+        offload_stream: torch.cuda.Stream,
+    ):
+        if not torch.is_tensor(device_tensor):
+            raise ValueError(f"Expect stensor but got {device_tensor}")
+        self.device_tensor: Optional[torch.Tensor] = device_tensor
+        self.cpu_tensor: Optional[torch.Tensor] = None
+        self.offload_stream = offload_stream
+        self.d2h_event: Optional[torch.cuda.Event] = None
+        self.h2d_event: Optional[torch.cuda.Event] = None
+        self.device: torch.device = device_tensor.device
+
+    def copy_d2h_async(self) -> None:
+        current_stream = torch.cuda.current_stream()
+        self.offload_stream.wait_stream(current_stream)
+        with torch.cuda.stream(self.offload_stream):
+            self.cpu_tensor = self.device_tensor.to(
+                torch.device("cpu"), non_blocking=True
+            )
+            self.d2h_event = self.offload_stream.record_event()
+
+    def copy_h2d_async(self) -> None:
+        if self.device_tensor is not None:
+            return
+        assert self.cpu_tensor is not None
+        self.device_tensor = torch.empty_like(self.cpu_tensor, device=self.device)
+        self.offload_stream.wait_stream(torch.cuda.current_stream())
+        with torch.cuda.stream(self.offload_stream):
+            self.device_tensor.copy_(self.cpu_tensor, non_blocking=True)
+            self.h2d_event = self.offload_stream.record_event()
+
+    def wait_for_d2h(self):
+        if self.d2h_event:
+            torch.cuda.current_stream().wait_event(self.d2h_event)
+        self.device_tensor = None
+
+    def wait_for_h2d(self):
+        if self.h2d_event:
+            torch.cuda.current_stream().wait_event(self.h2d_event)
+        self.cpu_tensor = None
+
+
+class offload_to_cpu(saved_tensors_hooks):
+    """
+    This represents a saved tensors hooks context that offloads activations to
+    CPU in forward and un-offloads them from CPU in backward.
+
+    In forward, the D2H copy is always async. Device memory is freed when the
+    user calls :meth:`wait_for_d2h`, which should be done after the compute
+    with which to overlap has been issued.
+
+    In backward, the H2D copy defaults to sync. However, the user may call
+    :meth:`copy_h2d_async` to issue the H2D copy as async before the compute
+    with which to overlap has been issued. When the activation is used in
+    backward, we will wait for that H2D copy without user intervention.
+
+    The D2H and H2D copies always used pinned memory, so the user should take
+    care to ensure sufficient CPU RAM to be pinned. Otherwise the program can
+    become slow or freeze. The first few iterations will be much slower due to
+    repeated ``cudaHostAlloc`` calls to warmup the CPU caching allocator.
+    """
+
+    def __init__(self, offload_stream: torch.cuda.Stream):
+        self.handle_key_to_handle: Dict[HandleKey, Handle] = {}
```

Comment:
to-do: maybe expose a property like `def handles(self) -> List[Handle]` since the `handle_key` is only needed internally and does not need to be exposed to user